File: Code/Arrays_Bit_Manipulation/Subarray_Sum_Divisible_By_K.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        unordered_map<int, int> remainderCount;
        remainderCount[0] = 1;

        int prefixSum = 0, count = 0;

        for (int num : nums) {
            prefixSum += num;
            int remainder = ((prefixSum % k) + k) % k;

            if (remainderCount.count(remainder)) {
                count += remainderCount[remainder];
            }
            remainderCount[remainder]++;
        }

        return count;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {4, 5, 0, -2, -3, 1};
    int k = 5;

    int result = sol.subarraysDivByK(nums, k);
    cout << "Number of subarrays divisible by " << k << " is: " << result << endl;

    return 0;
}

File: Code/Arrays_Bit_Manipulation/Subarray_Sum_Divisible_By_K.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int SubarraysDivByK(int[] nums, int k) {
        var remainderCount = new Dictionary<int, int>();
        remainderCount[0] = 1;

        int prefixSum = 0, count = 0;

        foreach (int num in nums) {
            prefixSum += num;
            int remainder = ((prefixSum % k) + k) % k;

            if (remainderCount.ContainsKey(remainder)) {
                count += remainderCount[remainder];
                remainderCount[remainder]++;
            } else {
                remainderCount[remainder] = 1;
            }
        }

        return count;
    }
}

public class Program {
    public static void Main(string[] args) {
        var sol = new Solution();
        int[] nums = { 4, 5, 0, -2, -3, 1 };
        int k = 5;

        int result = sol.SubarraysDivByK(nums, k);
        Console.WriteLine($"Number of subarrays divisible by {k} is: {result}");
    }
}
